src/lib/importSystem/record.ts

```diff
@@ -0,0 +1,562 @@
+    availableFiles
+      .filter((availableFile) => !currentImportFiles
+        .some((currentImportFile) => (
+          importId === currentImportFile.importId
+          && availableFile.fileInfo.path === currentImportFile.fileInfo.path
+          && availableFile.fileInfo.name === currentImportFile.fileInfo.name
+        ))),
+    // Matched files are those that are already recorded in the database
+    availableFiles
+      .filter((availableFile) => currentImportFiles
+        .some((currentImportFile) => (
+          importId === currentImportFile.importId
+          && availableFile.fileInfo.path === currentImportFile.fileInfo.path
+          && availableFile.fileInfo.name === currentImportFile.fileInfo.name
+        ))),
+    // Removed files are those that were recorded in the database but are no longer available
+    currentImportFiles
+      .filter((currentImportFile) => !availableFiles
+        .some((availableFile) => (
+          importId === currentImportFile.importId
+          && availableFile.fileInfo.path === currentImportFile.fileInfo.path
+          && availableFile.fileInfo.name === currentImportFile.fileInfo.name
+        ))),
+  ];
+
+  return Promise.all([
+    // Create new files in the database if there are any
+    (newfiles.length > 0
+      ? ImportFile.bulkCreate(
+        newfiles.map((newFile) => ({
+          importId,
+          ftpFileInfo: newFile.fileInfo,
+          status: IMPORT_STATUSES.IDENTIFIED,
+        })),
+        { independentHooks: true },
+      )
+      : Promise.resolve()),
+    // Update matched files in the database if there are any
+    ...(matchedFiles.length > 0
+      ? matchedFiles.map(async (matchedFile) => ImportFile.update(
+        {
+          importId,
+          ftpFileInfo: matchedFile.fileInfo,
+        },
+        {
+          where: {
+            importId,
+            [Op.and]: [
+              Sequelize.literal(`"ftpFileInfo" -> 'path' = '${matchedFile.fileInfo.path}'`),
+              Sequelize.literal(`"ftpFileInfo" -> 'name' = '${matchedFile.fileInfo.name}'`),
+            ],
+          },
+        },
+      ))
+      : []),
+    // Delete removed files from the database if there are any
+    (removedFiles.length > 0
+      ? ImportFile.destroy({
+        where: {
+          importId,
+          id: removedFiles.map(({ id }) => id),
+        },
+        independentHooks: true,
+      })
+      : Promise.resolve()),
+  ]);
+};
+
+const recordAvailableDataFiles = async (
+  importFileId: number,
+  availableFiles: ZipFileInfo[],
+) => {
+  const currentImportDataFiles: {
+    id: number,
+    importFileId: number,
+    fileInfo: ZipFileInfo,
+  }[] = await ImportDataFile.findAll({
+    attributes: [
+      'id',
+      'importFileId',
+      ['zipFileInfo', 'fileInfo'],
+    ],
+    where: {
+      importFileId,
+    },
+    raw: true,
+  });
+
+  // Separate the available files into new, matched, and removed files
+  const [
```

Comment:
this logic looks identical to the above, can we dry this up a little?

Reply:
updated to new format